code/game/objects/items/RCD.dm

```diff
@@ -1133,53 +1164,96 @@ GLOBAL_VAR_INIT(icon_holographic_window, init_holographic_window())
 	return ..()
 
 /obj/item/construction/plumbing/attack_self(mob/user)
-	..()
-	if(!choices.len)
-		for(var/obj/machinery/plumbing/plumbing_type as anything in plumbing_design_types)
-			choices += list(initial(plumbing_type.name) = image(initial(plumbing_type.icon), icon_state = initial(plumbing_type.icon_state)))
-			name_to_type[initial(plumbing_type.name)] = plumbing_type
-			machinery_data["cost"][plumbing_type] = plumbing_design_types[plumbing_type]
+	ui_interact(user)
 
-	// Update duct icon
-	var/image/duct_image = choices["fluid duct"]
-	duct_image.color = current_color
+/obj/item/construction/plumbing/examine(mob/user)
+	. = ..()
+	. += "You can scroll your mouse wheel to change the piping layer."
+	. += "You can right click a fluid duct to set the Plumbing RPD to its color and layer."
+
+/obj/item/construction/plumbing/ui_interact(mob/user, datum/tgui/ui)
+	ui = SStgui.try_update_ui(user, src, ui)
+	if(!ui)
+		ui = new(user, src, "PlumbingService", name)
+		ui.open()
+
+/obj/item/construction/plumbing/ui_assets(mob/user)
+	return list(
+		get_asset_datum(/datum/asset/spritesheet/plumbing),
+	)
 
-	var/choice = show_radial_menu(user, src, choices, custom_check = CALLBACK(src, PROC_REF(check_menu), user), require_near = TRUE, tooltips = TRUE)
-	if(!choice || !check_menu(user))
+/obj/item/construction/plumbing/ui_static_data(mob/user)
+	var/list/data = list("paint_colors" = GLOB.pipe_paint_colors)
+	return data
+
+/obj/item/construction/plumbing/ui_data(mob/user)
+	var/obj/machinery/b = blueprint
+	//map string layer to number layer . calculations don't look nice but better than creating an list in memory to do the mapping
+	var/layer = GLOB.plumbing_layers[current_layer]
+	if(layer == 16)   //max bit flag means layer 5
+		layer = 5
+	else if(layer > 2) //for bit flags 4 & 8 we do bitflag/4 + 2 . for example 4/4 + 2 = 3[layer 3], 8/4 +2 = 4[layer 4]
+		layer = (layer / 4) + 2
+
+	var/list/data = list()
+	data["piping_layer"] = layer
+	data["selected_color"] = current_color
+	data["icon"] = initial(b.icon_state)
+
+	var/index = 1
+	var/list/category_list = list()
+	var/selected = initial(b.name)
+	var/category = ""
+
+	for(var/obj/machinery/recipe as anything in plumbing_design_types)
+		category = ""  //find category of each design type
+		if(ispath(recipe , /obj/machinery/plumbing))
+			var/obj/machinery/plumbing/p = recipe
+			category = initial(p.category)
+		else if(ispath(recipe , /obj/machinery/duct))
+			var/obj/machinery/duct/p = recipe
+			category = initial(p.category)
+		else
+			var/obj/machinery/iv_drip/plumbing/p = recipe
+			category = initial(p.category)
+
+		var/list/item = list()
+		item["index"] = index++
```

Comment:
Don't use the output value of increment operators, it makes the code more confusing